SmartContracts.NET/StrongForce.Core/Permissions/AccessControlList.cs

```diff
@@ -1,92 +1,66 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using StrongForce.Core;
 
 namespace StrongForce.Core.Permissions
 {
 	public class AccessControlList
 	{
-		public AccessControlList(IDictionary<Permission, IDictionary<Address, HashSet<Address>>> initialWildCards)
-			=> this.PermissionsoToReceiverToSenders = initialWildCards;
+		public AccessControlList(ISet<Permission> initialPermissions)
+		{
+			this.Permissions = initialPermissions;
+		}
 
 		public AccessControlList()
-			: this(new SortedDictionary<Permission, IDictionary<Address, HashSet<Address>>>())
+			: this(new SortedSet<Permission>())
 		{
 		}
 
-		public IDictionary<Permission, IDictionary<Address, HashSet<Address>>> PermissionsoToReceiverToSenders { get; set; }
+		public ISet<Permission> Permissions { get; set; }
 
-		public IEnumerable<Address> GetPermittedAddresses(Permission permission, Address target)
+		public IEnumerable<Permission> FilterPermissions(Permission pattern)
 		{
-			var sendersToReceivers = this.PermissionsoToReceiverToSenders[permission];
-
-			return this.PermissionsoToReceiverToSenders[permission][target];
+			return this.Permissions.Where(permission => permission.Match(pattern));
 		}
 
-		public bool HasPermission(Address sender, Permission permission, Address receiver)
+		public bool HasPermission(Permission pattern)
 		{
-			if (sender == null || permission == null || receiver == null || !this.PermissionsoToReceiverToSenders.ContainsKey(permission))
-			{
-				return false;
-			}
-
-			bool check1 = false, check2 = false;
-			this.PermissionsoToReceiverToSenders[permission].TryGetValue(receiver, out HashSet<Address> perm1);
-			this.PermissionsoToReceiverToSenders[permission].TryGetValue(Address.Null, out HashSet<Address> perm2);
-			if (perm1 != null)
-			{
-				check1 = perm1.Contains(sender) || perm1.Contains(Address.Null);
-			}
-
-			if (perm2 != null)
-			{
-				check2 = perm2.Contains(sender) || perm2.Contains(Address.Null);
-			}
-
-			return check1 || check2;
+			return this.Permissions.Any(permission => permission.Match(pattern));
```

Comment:
We should use ``permission.Equals(other)`` here, because the Contract class currently is not using this method, and it's not breaking things, but if a derived contract uses it, it will return true for a Permission which has everything to null except the type.

Reply:
We are using `Match` here, as this allows for having permissions for any address or any action type.